### DISCUSSION TITLE
Readonly postgres user unable to fetch tables

### DISCUSSION THREAD

**Discussion by imakiro** [Budibase/budibase#9231] [Help] [answered]
Hi,

It seems that a read-only user cannot "fecth tables" when being used to connect to a postgress DB as a data source.

```
Error fetching tables
No primary key constraint found for the following: 
*here lies all the DB tables* 
```

I have found two errors relating to this message, one relating to views and one relating to primary keys :
https://github.com/Budibase/budibase/issues/6216
https://github.com/Budibase/budibase/discussions/3840

The affected tables are tables, not views, and they have a primary key, as many of them are django related and this framework supports CRUD on this DB.

So the tables are indeed found, but not the contraints?
Considering that using a read write user works fine, is there any specific permissions that I would've missed in the docs that would be required ?

**Comment by melohagan** [comment]
Hey @imakiro 

Could you please share the schema of your database tables. I.e. the CREATE scripts for your Postgres tables

**Reply by wilsonaaronj** [reply]
I'm having this same issue.   My tables have primary key constrains and I'd really like to test with read-only on my sensitive data for now.  Any simple suggestions to how to achieve this? Thanks

**Comment by travismiller** [comment]
I've run into this also.

Seems the relevant Postgres SQL for getting primary keys is here:
https://github.com/Budibase/budibase/blob/562ee514acf8f6ae9653d454a49cf0d4e6f88463/packages/server/src/integrations/postgres.ts#L126-L134

Which looks proper.

Testing this query directly on the DB I've run into demonstrates that the "read only" credential returns no rows with that query. Though that credential is granted SELECT on several tables. I'm not sure what I don't know here.

Note that this other query on `information_schema` does return rows as expected:
https://github.com/Budibase/budibase/blob/562ee514acf8f6ae9653d454a49cf0d4e6f88463/packages/server/src/integrations/postgres.ts#L167

For reference, the issue in question for me is on Heroku Postgres with a credential add-on with the read only grants. See [Heroku addon credential comment on StackOverflow].

A different query, adapted from [primary keys query comment on StackOverflow] does give me primary key rows with the read-only credential.

```sql
SELECT pg_namespace.nspname "table_schema"
     , pg_class.relname "table_name"
     , pg_attribute.attname "primary_key"
     , format_type(pg_attribute.atttypid, pg_attribute.atttypmod) "data_type"
FROM pg_class
JOIN pg_index ON pg_class.oid = pg_index.indrelid AND pg_index.indisprimary
JOIN pg_attribute ON pg_attribute.attrelid = pg_class.oid AND pg_attribute.attnum = ANY(pg_index.indkey)
JOIN pg_namespace ON pg_namespace.oid = pg_class.relnamespace
WHERE pg_namespace.nspname = :schema
;
```

[Heroku addon credential comment on StackOverflow]: https://stackoverflow.com/a/53785008/1596013
[primary keys query comment on StackOverflow]: https://stackoverflow.com/a/20537829/1596013

**Reply by adrinr** [reply, marked as answer]
This is being resolved in here (https://github.com/Budibase/budibase/issues/10208), thank you for the detailed information
